### DISCUSSION TITLE
Definition of base generalized acceleration and centroidal momentum matrix time variation

### DISCUSSION THREAD

**Discussion by Yukai-Gong** [stack-of-tasks/pinocchio#1829] [Q&A] [answered]
Hi,

I am pretty sure the generalized velocity for floating base model in Pinocchio is always `[lin_vel_in_base_frame, ang_vel_in_base_frame, joint_vel]`. However I am not sure about the generalized acceleration.

- In [1656](https://github.com/stack-of-tasks/pinocchio/issues/1656). The acceleration is `[lin_acc_in_base_frame, ang_acc_in_base_frame, joint_acc]`.

- However, in the documentation for [dAg](https://gepettoweb.laas.fr/doc/stack-of-tasks/pinocchio/master/doxygen-html/structpinocchio_1_1DataTpl.html#ac353f76ffcda5c59657f0f8ce57c55bd), it looks like the q_ddot is defined as `[d/dt(lin_vel_in_base_frame), d/dt(ang_vel_in_base_frame), joint_acc]`

I would like to confirm the followings:

1. When we are using dhg = Ag * q_ddot + dAg * q_dot, we should plug in `[d/dt(lin_vel_in_base_frame), d/dt(ang_vel_in_base_frame), joint_acc]` (And the same for many other jacobian time variations)

2. The inertia matrix (data.M) is computed for `[lin_acc_in_base_frame, ang_acc_in_base_frame, joint_acc]`

Thanks!

**Comment by jcarpent** [comment, marked as answer]
To quickly answer your questions, it is important to know that for joints in general, velocities, forces and their variants are all expressed in the frame of the joint itself. It follows the rules from Featherstone in his seminal book.
The same story applies to the joint velocity, acceleration or torque vectors, where the quantities are expressed in the local frame of the joints.

Hope it will answer your question.

**Reply by Yukai-Gong** [reply]
Thanks for the reply. I understand that joint quantities are expressed in joint frames. But I am not sure about the floating base. I think `base_lin_acc_in_base_frame` and `d/dt(base_lin_vel_in_base_frame)` are different variables. And it looks like the former is used in some places and the latter is used in other places. I would like to confirm if it is true. Thanks.

**Reply by jcarpent** [reply]
Yes, you can look at Featherstone book for clarification (formula 2.50). You should consider that you have spacial acceleration and classic acceleration.

**Reply by Yukai-Gong** [reply]
Thank you! So the Inerita Matrix, Coriolis Matrix, Gravity Vector, and Nonlinear Effect provided by pinocchio are all computed for the spatial acceleration instead of the classical acceleration.

**Reply by jcarpent** [reply]
Yes, indeed.
